### DISCUSSION TITLE
Introducing New Design Blueprint  Using Third Parties - Stage 1 Networking

### DISCUSSION THREAD

**Discussion by hassannmoussaa** [GoogleCloudPlatform/cloud-foundation-fabric#2159] [General]
# Idea Proposal

As the need for landing zone deployment on GCP is growing, and automation is always needed, and from the fact the using third parties from Google Partners are a must to fulfil the customer needs, new blueprint from the recommended designs by third parties become needed. From my experience at a Professional Services engineer at Google Premier Partner, we are requested every day to deploy hub-spoke models using third parties tools such as Palo Alto VM_Series , fortigate , f5 , .... 

Based on the above mentioning, i would like to suggest ( and would love to work on this as contribution) to have new blueprint on the networking stage that introduce more dynamic and production-ready terraform templates and i'm listing down one of the suggestion that we can work om

## Dedicated NVA - Third Party Hub & Spoke Model 

As per palo alto and google official documentation which you can refer to using the links in the references, the recommended design for palo alto on gcp now become either having dedicated pair of VMs for inbound and outbound, or single pair of VMs for both direction, in my today issue im suggesting having a terraform ready blueprint that deploy the following production-ready hub-spoke dynamic model in the following design.

![BluePrint- Palo Alto On GCP - Dedicated](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/assets/47701743/29cb298a-76f7-4830-9e12-a67a54d84a5a)

## Architecture Explanation 

The above design is a hub and spoke design based on a hub project that include a trusted and untrusted zone where traffic flow move from untrusted to trusted passing by load balancer with the first set of  VM-Series ( Palo Alto) backend for inspection and threat detection.

Because hybrid connectivity is always a need, a VPC for VPN is needed, where this VPC is peered to the trusted and because transitive peering is not allowed this traffic will pass by the second set of vm-series for inspection

Outbound traffic initiating from the VMs are inspected by another set of palo alto to eliminate bandwidth issue and for better visibility.

Trusted VPC is peered with a set of prod and non prod shared VPCs that share their network with service project that contain the actual workload.

## Proposed Work 

I would like to contribute and to split the following issue and feature into multiple issues to make easier for maintainer to conduct the code review and validation.

In my proposal , i would suggest to work on something similar to the available blueprint in stage 2, where we can have a fully dynamic terraform model that deploy this well needed architecture on GCP following the design principles followed  by fabric fast modules and following the same structure.

## Next Steps

i would like to hear from this repository maintainer about this idea in order to agree on it, thus i have some ready modules that i can work on to match fabric fast design principles and contribute to the repository with this new design in order to have it ready in the next releases .

Moreover from the fact that issues are a way to introduce feature and issues should be simple and direct, i would like to split this feature into multiple smaller issues in next work in order to merge this change step by step 

As a next step , I'm planning to involve other recommended designs such as :

- Shared set of palo alto hub and spoke model 
- FortiGate recommended design 

And other recommended design that we feel they are very requested nowadays in order to enrich current fabric fast foundation modules with more production-ready and user-needed designs and architecture 

## Conclusion

i would like to know more from this page maintainer about my thoughts and im always ready to help and contribute

**Comment by ludoo** [comment]
@sruffilli can you chime in here?

**Comment by hassannmoussaa** [comment]
Appreciate your thoughts, @sruffilli

**Comment by hassannmoussaa** [comment]
A gentle reminder

**Comment by sruffilli** [comment]
Will TAL by EOD :)

**Comment by sruffilli** [comment]
Hey @hassannmoussaa, thanks for kicking off the discussion.
While I understand and appreciate the need for additional blueprints implementing the reference designs for the different vendors, there's two main things I'm concerned about:

*  The proliferation of networking stages is (already) a **maintainability burden**. Unfortunately every time a module updates its interface, or as we come up with a novel implementation (e.g. dns policies for PGA), 5 different network stages must be painstakingly updated and aligned with each other. Coming up with multiple new stages would only aggravate the problem.
* **Lack of testability** due to an explicit dependency to appliances requiring a license to be deployed. While this could be alleviated by deploying `simple-nva` placeholders, doing so would IMHO result in a futile exercise given a good 95% (100%?) of the rest of the infrastructure would be vendor agnostic.

A low hanging fruit could be adding additional VPCs (at least one for management and one for heartbeat, which is pretty common across vendors) to the existing `2-networking-c-nva`.

Something that could be done on top of that would be re-architecting (as/if required) the current `2-networking-c-nva` stage to parcel out everything that is related to NVAs to clearly identified `nva-*.tf` files, and implement (in a separate "3p-plugins" folder?) drop-in terraform files for vendor-specific solutions. 

@ludoo @juliocc wdyt?

**Comment by hassannmoussaa** [comment]
@sruffilli 

Thank you for your insightful response. While i totally agree with your mention, allow me to suggest some potential workaround from my side. 

- While adding stages will be a maintenance nightmare, we can reduce changable features like dns,... into modules with default values, this could provide some facilitation to add more stages while editing the basic features from module default values.

- Im wondering if we can use a gcp free trial license (payg free trial) as testing material for almost all well-known NVA

- Editing basic model of NVA would be a great solution as the typical NVA deployment nowadays is using these 4 VPCs, moreover introducing a fully production-ready blueprint stages will empower end-user with architecture that i can see they are well implemented and requested today on GCP  ( like 1 layer palo alto model , 3 layers fortigate -- F5 -- Palo Alto ) and so on 

I would love to hear from all of you your thoughts while really intersted to contribute to what we could agree on. 

🚀

**Reply by juliocc** [reply]
> While adding stages will be a maintenance nightmare, we can reduce changable features like dns,... into modules with default values, this could provide some facilitation to add more stages while editing the basic features from module default values.

This will require calling modules within modules, which goes directly against our [design principles](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md#the-zen-of-fabric).

> Im wondering if we can use a gcp free trial license (payg free trial) as testing material for almost all well-known NVA

The problem is not about licensing/cost. We're talking about unit or e2e testing.

> Something that could be done on top of that would be re-architecting (as/if required) the current 2-networking-c-nva stage to parcel out everything that is related to NVAs to clearly identified nva-*.tf files, and implement (in a separate "3p-plugins" folder?) drop-in terraform files for vendor-specific solutions.

I like the plugin idea but I'm not sure how far we can push it. If it works out, this would be a reasonable compromise IMO.

@hassannmoussaa What is your goal having this in the core Fabric/FAST repo? You could potentially host this network stage in a separate repo (and we could link your repo from the networking READMEs)

**Reply by hassannmoussaa** [reply]
@juliocc Thank you for mentioning this , 

From my basic experience, i can see that fabric fast is being used officially for deploying landing zone architecture proposed by the PSO team

While i can see that the customers today care about migrations and IAAS and still we are step behind the fully modernized architecture,  security is a concern .


Basic on that, the customers are asking more for NVA that could secure Iaas as per the shared responsibility model.

Having it on the fabric fast could facilitate this mission for  the  system integrator. This could provide it more credibility. 

While i can see this, This will add a layer of complexity over the basic designs and implementation.

**Comment by ludoo** [comment]
Plus one on the proposal to add the missing bits to the current nva stages, this is painless and actually pretty useful.

Encapsulating functionality specific to network stages is something we will never do, as we always strived very hard to reduce the number of layers.

**Comment by hassannmoussaa** [comment]
@ludoo Thank you for your clarification 

So one thing is we can proceed to edit the current nva stage to include 4 VPCs and almost peering with shared VPCs , while we can make the number of NVA dynamic for dedicated inboud/outbound and HA options ( if you agree on this proposal ) 

Can you provide your thoughts on having a 3-plugin folder for deploying vendor specific NVA along with their basic configuration (HA , interfaces ,... ) , do you see that a good addition ? And how shall we consider it on this change 

Thank you for your time

**Reply by sruffilli** [reply]
> [...] edit the current nva stage to include 4 VPCs and almost peering with shared VPCs , while we can make the number of NVA dynamic for dedicated inboud/outbound and HA options (if you agree on this proposal).

Yay for the two additional VPCs - even though I believe that the "dedicated NIC for management" pattern is a relic from the onprem world and only makes sense if there's a dedicated VPN/VLAN attachment for it, which is almost never the case.
Nay for inbound/outbound, which to be honest I'm not really seeing in my day-to-day.
Wdym with "almost peering"?

> Can you provide your thoughts on having a 3-plugin folder for deploying vendor specific NVA along with their basic configuration (HA , interfaces ,... ) , do you see that a good addition ? And how shall we consider it on this change

I would personally like to see that happening, and would be happy to look into what changes are required to `2-networking-c-nva` to make it happen, but I have doubts about that folder existing within the Fabric repo, due to the inability to do e2e testing, potentially resulting in stale code. Either an external repo (as @juliocc recommends) or a dedicated folder with clear per-contrib ownership would IMHO be a good compromise.

**Reply by ludoo** [reply]
+1 but I would avoid a dedicated folder, it would still introduce substantial maintenance overhead...

**Reply by hassannmoussaa** [reply]
@ludoo @juliocc @sruffilli 

What about proceeding with the esit for the additional VPC in the nva stage and editing the documentation 


While working on an external repository for deploying vendor specific deployment such as vm-series , fortigate,..

I have one question here , if we want to work on an external repo and the deployment model needs editing for vpc or require additional VPCs like having multiple NVA how we could manage that or that is not a requirement at the moment

Moreover, could we consider having peering with shared VPC for the current nva stage

Appreciate your thoughts

**Reply by sruffilli** [reply]
> What about proceeding with the esit for the additional VPC in the nva stage and editing the documentation

Sounds good, go ahead and send a PR!

> I have one question here , if we want to work on an external repo and the deployment model needs editing for vpc or require additional VPCs like having multiple NVA how we could manage that or that is not a requirement at the moment

I suspect all of that can be parceled out to those "plugins". But I guess it'll be much more clear as you'll start developing the first plugin.

> Moreover, could we consider having peering with shared VPC for the current nva stage

Can you please give me more details here?

Thanks!

**Reply by hassannmoussaa** [reply]
@sruffilli i meant changing the type of the peered vpc in the initial design ( the two peered VPC) to shared VPC not normal VPC so other service project can consum network 

Ill move forward with the agreed PR 

Thank you

**Reply by hassannmoussaa** [reply]
@sruffilli @ludoo one more request, can you please share with me the architecture included in nva documentation as lucid chart document if that is possible 

Appreciate it , below my email 

hassannmoussaa@gmail.com

**Reply by sruffilli** [reply]
> @sruffilli i meant changing the type of the peered vpc in the initial design ( the two peered VPC) to shared VPC not normal VPC so other service project can consum network

Ah sure

> @sruffilli @ludoo one more request, can you please share with me the architecture included in nva documentation as lucid chart document if that is possible

You meant Google Slide :)? On its way.

**Comment by hassannmoussaa** [comment]
PR Opened

https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/pull/2180

**Comment by hassannmoussaa** [comment]
I’d like to reopen this topic as an enhancement. It was proposed some time ago, but I wasn’t able to address it back then.

Is there still interest in moving this forward? If so, I can draft a PR and proceed.
